File: packages/python/examples/demos/core_runtime_safety_demo.py
"""1.14.2 mini-demo: offline fallback + safe tools + structured handoff (Python).

Justifies the patch release alongside C++ handoffkit::core packaging and the
matching JS/Rust demos. No network required.
"""

from __future__ import annotations

from typing import Any

from handoffkit import Agent, HandoffProtocol, Team, __version__
from handoffkit.providers.echo_provider import EchoProvider
from handoffkit.providers.fallback import FallbackProvider
from handoffkit.tools import extract_keywords, summarize_text


class _FailingProvider:
    """Primary that always fails so fallback is exercised."""

    model = "primary-fail"

    def generate(self, prompt: str, **kwargs: Any) -> str:
        raise RuntimeError("primary down (demo)")


def main() -> None:
    print(f"handoffkit {__version__} — core_runtime_safety_demo (Python)")
    print("theme: FallbackProvider + offline text tools + hybrid_state handoff\n")

    # 1) Fallback: fail primary → echo secondary
    fb = FallbackProvider(
        [_FailingProvider(), EchoProvider(model="secondary-echo")],
        model="fallback-chain",
    )
    answer = fb.generate("Ship a minimal offline multi-agent handoff.")
    assert "EchoProvider" in answer or "Summary" in answer
    print("[fallback] secondary echo OK")
    print(answer[:240].rstrip() + ("...\n" if len(answer) > 240 else "\n"))

    # 2) Offline text tools (not shell / not run_command)
    summary = summarize_text(
        "HandoffKit ships installable C++ core plus Python and JS parity demos."
    )
    keys = extract_keywords(summary)
    print(f"[tools] summarize_text chars={len(summary)}")
    print(f"[tools] extract_keywords sample={keys[:5]}")

    # 3) Tiny dual-agent team with structured protocol
    protocol = HandoffProtocol(mode="hybrid_state")
    team = Team(
        agents=[
            Agent(
                name="Architect",
                role="Plan offline-safe multi-runtime handoffs.",
                provider=EchoProvider(model="arch"),
            ),
            Agent(
                name="Coder",
                role="Implement installable core surfaces.",
                provider=EchoProvider(model="coder"),
            ),
        ],
        protocol=protocol,
    )
    result = team.run("Document handoffkit::core install for C++ and parity demos for Python/JS.")
    print(f"[team] handoffs={len(result.handoffs)} final_output_chars={len(result.final_output)}")
    print("\ncore_runtime_safety_demo OK")


if __name__ == "__main__":
    main()

File: packages/js/core/examples/core_runtime_safety_demo.js
/**
 * 1.14.2 mini-demo: offline fallback + structured handoff (JavaScript).
 * Mirrors Python examples/demos/core_runtime_safety_demo.py and C++ consumer_core.
 * No network required.
 */
import {
  Agent,
  EchoProvider,
  FailingProvider,
  FallbackProvider,
  HandoffProtocol,
  Team,
} from "../src/index.js";

function main() {
  console.log("handoffkit JS — core_runtime_safety_demo (1.14.2)");
  console.log("theme: FallbackProvider + Team hybrid_state handoff\n");

  const fb = new FallbackProvider({
    providers: [
      new FailingProvider({ model: "primary-fail", message: "primary down (demo)" }),
      new EchoProvider({ model: "secondary-echo" }),
    ],
    model: "fallback-chain",
  });
  const answer = fb.generate("Ship a minimal offline multi-agent handoff.");
  if (!String(answer).includes("secondary-echo") && !String(answer).includes("Echo")) {
    throw new Error("fallback did not return echo secondary");
  }
  console.log("[fallback] selectedModel=", fb.selectedModel);
  console.log(String(answer).slice(0, 200), "...\n");

  const team = new Team({
    agents: [
      new Agent({
        name: "Architect",
        role: "Plan offline-safe multi-runtime handoffs.",
        provider: new EchoProvider({ model: "arch" }),
      }),
      new Agent({
        name: "Coder",
        role: "Implement installable core surfaces.",
        provider: new EchoProvider({ model: "coder" }),
      }),
    ],
    protocol: new HandoffProtocol({ mode: "hybrid_state" }),
  });

  const result = team.run("Document handoffkit::core install for C++ and parity demos for Python/JS.");
  console.log("[team] success=", result.success, "handoffs=", (result.handoffs || []).length);
  console.log("[team] final_output_chars=", String(result.finalOutput || "").length);
  console.log("\ncore_runtime_safety_demo OK");
}

main();
